### PR TITLE
feat(eslint-plugin): [no-confusing-void-expression] add an option to ignore void<->void

### LINKED ISSUE #8538: Enhancement: [no-confusing-void-expression] Add option to ignore void<->void
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-confusing-void-expression

### Description

Following from https://github.com/typescript-eslint/typescript-eslint/pull/8375#pullrequestreview-1897027584: we tried enabling `@typescript-eslint/no-confusing-void-expression` internally and very quickly found it to be inconvenient:

> Looking at the changes - TBH I don't really like this rule. I don't think any of these changes improve clarity or readability of code.
> 
> The biggest issue I see is that if the function is implicitly typed or explicitly marked as returning `void` the rule still requires you to change things.
> 
> For example this code should pass the rule, IMO:
> 
> ```ts
> function returnsVoid(): void {}
> 
> function test1(): void {
>   return returnsVoid(); // should be fine
> }
> const test2 = (): void => returnsVoid(); // should be fine
> ```
> 
> Because there's nothing confusing about that code - the expressions are all typed as `void` and the functions are typed as `void`.
> 
> Also changes like `() => expr` => `() => { expr }` actively reduce readability, IMO, esp when considering the above.

Agreed. @bradzacher and I chatted briefly, and Brad suggested:

1. Adding an option to ignore the `void` <-> `void` return case
2. Making that option the default in v8

Agreed! Filing this issue to add that option. We can file a followup if & when it gets in.

### Fail

```TypeScript
declare function returnsVoid(): void;

function outerReturnsVoid(): void {
  return returnsVoid();
}
```

### Pass

```TypeScript
declare function returnsString(): string;

function outerReturnsString(): string {
  return returnsString();
}
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by ronami**
## PR Checklist

- [x] Addresses an existing open issue: fixes #8538
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

This PR addresses #8538 and adds an option (namely `ignoreVoidReturningFunctions`) that allows returning `void` expressions from either:

- Functions explicitly annotated with `void` as one of their return types.
- Their contextual type includes `void`.

I enabled the rule back on `@typescript-eslint`'s monorepo (since it's currently disabled) with the option and had to fix one error that looks valid. Running the rule on `@typescript-eslint`'s monorepo helped me figure out the various edge cases of the rule (hopefully, I didn't miss a whole lot).

I added a "game plan" comment after seeing @kirkwaiblinger do this on several other rules, and it has been helpful to read through.

One open question I have is whether or not this should apply to async functions. I mean, if the following should be valid or not:

```ts
function returnsVoid(): void {}

async function foo(): Promise<void> {
  return returnsVoid();
}
```

Please let me know if this makes sense or if I missed anything.

Edit: Only when I finished most of the work I noticed the linked PR of https://github.com/typescript-eslint/typescript-eslint/pull/8632. I noticed I was missing a bunch of tests and added the ones I was missing, thanks @developer-bandi 🙏

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/no-confusing-void-expression.mdx [hunk 1]
```diff
@@ -119,6 +119,24 @@ function doSomething() {
 console.log(void alert('Hello, world!'));
 ```
 
+### `ignoreVoidReturningFunctions`
+
+Whether to ignore returns from functions with explicit `void` return types and functions with contextual `void` return types.
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/no-confusing-void-expression.mdx:124]: [Docs] I think most users won't understand what _"contextual"_  means. How does this strike you?

```suggestion
Whether to ignore returns from functions with `void` return types when inside a function with a `void` return type.
```

File: packages/eslint-plugin/docs/rules/no-confusing-void-expression.mdx [hunk 2]
```diff
@@ -119,6 +119,24 @@ function doSomething() {
...
+
+It may be preferable to allow functions that explicitly return `void` to return `void` expressions. This exception will also apply to functions with an implicit contextual return type of `void`.
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/no-confusing-void-expression.mdx:126]: [Docs] It's good to be explicit on _why_ something is or isn't preferable:

```suggestion
Some projects prefer allowing functions that explicitly return `void` to return `void` expressions. Doing so allows more writing more succinct functions.

:::note
This is technically risky as the `void`-returning function might actually be returning a value not seen by the type system.
:::
```

File: packages/eslint-plugin/src/util/getParentFunctionNode.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/util/getParentFunctionNode.ts:1]: [Praise] I like the reuse here 🙂 nice.

File: packages/eslint-plugin/src/rules/no-confusing-void-expression.ts [hunk 3]
```diff
@@ -388,5 +415,56 @@ export default createRule<Options, MessageId>({
...
+
+        return tsutils
+          .unionTypeParts(returnType)
+          .some(tsutils.isIntrinsicVoidType);
+      }
+
+      if (
+        ts.isFunctionExpression(functionTSNode) ||
+        ts.isArrowFunction(functionTSNode)
+      ) {
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-confusing-void-expression.ts:457]: [Style] Nit: a little less code to read & run here, as `getContextualType` is just looking for expressions:

```suggestion
      if (ts.isExpression(functionTSNode)) {
```

File: packages/eslint-plugin/src/rules/no-confusing-void-expression.ts [hunk 4]
```diff
@@ -388,5 +415,56 @@ export default createRule<Options, MessageId>({
...
+      //   - Return based on whether the contextual type includes `void` or not
+
+      const functionTSNode = services.esTreeNodeToTSNodeMap.get(functionNode);
+
+      if (functionTSNode.type) {
+        const returnType = checker.getTypeFromTypeNode(functionTSNode.type);
+
+        return tsutils
+          .unionTypeParts(returnType)
+          .some(tsutils.isIntrinsicVoidType);
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-confusing-void-expression.ts:451]: [Testing] No tests fail if this is changed to `.every`. I think that means there's some logic being missed?

File: packages/eslint-plugin/src/rules/no-confusing-void-expression.ts [hunk 5]
```diff
@@ -388,5 +415,56 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function isFunctionReturnTypeIncludesVoid(functionType: ts.Type): boolean {
+      const callSignatures = tsutils.getCallSignaturesOfType(functionType);
+
+      return callSignatures.some(signature => {
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-confusing-void-expression.ts:422]: [Testing] No tests fail if this is changed to `.every`. I think that means there's some logic being missed?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Looks like a great start to me! Mostly requesting changes on docs & testing, the game plan seems very reasonable. 🏈

File: packages/eslint-plugin/docs/rules/no-confusing-void-expression.mdx (hunk 1 above)
Comment (ronami) [reply packages/eslint-plugin/docs/rules/no-confusing-void-expression.mdx:124]: Yes! This is a lot better, thanks ❤️

File: packages/eslint-plugin/docs/rules/no-confusing-void-expression.mdx (hunk 2 above)
Comment (ronami) [reply packages/eslint-plugin/docs/rules/no-confusing-void-expression.mdx:126]: Can't agree more on this, thanks 👍

File: packages/eslint-plugin/src/rules/no-confusing-void-expression.ts (hunk 3 above)
Comment (ronami) [reply packages/eslint-plugin/src/rules/no-confusing-void-expression.ts:459]: Nice.

Comment (ronami) [reply packages/eslint-plugin/src/rules/no-confusing-void-expression.ts]: I've seen [this done here](https://github.com/typescript-eslint/typescript-eslint/blob/ef5df2f83bc20810f5b33ce31f691a8766990977/packages/eslint-plugin/src/rules/no-unsafe-return.ts#L104-L105), would it make sense to send a tiny PR to shorten it in a similar way? Looks like it works out locally.

File: packages/eslint-plugin/src/rules/no-confusing-void-expression.ts (hunk 5 above)
Comment (ronami) [reply packages/eslint-plugin/src/rules/no-confusing-void-expression.ts]: @JoshuaKGoldberg is it possible for a function expression (arrow or standard) to have multiple call signatures (other than function overloading)? I've currently changed this part of the code to assume only one call signature is available.

Comment (ronami) [reply packages/eslint-plugin/src/rules/no-confusing-void-expression.ts:424]: Okay, I figured this out and added the necessary test. 👍

**Comment by JoshuaKGoldberg** [comment]
@ronami is this ready for re-review?

We normally wait to be explicitly re-requested, so we're holding off posting more comments for now.

**Comment by ronami** [comment]
> @ronami is this ready for re-review?
> 
> We normally wait to be explicitly re-requested, so we're holding off posting more comments for now.

Oh, I wasn't aware of this. Yes! It is ready for review.

**Comment by JoshuaKGoldberg** [comment]
Great thanks!

In the future, you can use the actual GitHub feature: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews#re-requesting-a-review. That'll put it back on our queue. We have automations around it to remove the `awaiting response` label.

File: packages/eslint-plugin/src/rules/no-confusing-void-expression.ts (hunk 3 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/no-confusing-void-expression.ts]: If you want, I'd 👍 it!

**Review by JoshuaKGoldberg — APPROVED** [review]
Looks great, thanks for all the changes!

Renee Young hand-waving saying "Glorious!"

**Pull request merged**